src/test/java/org/broadinstitute/dsm/util/tools/TbosUPSKitTool.java

```diff
@@ -0,0 +1,510 @@
+                    }
+                    decideWhatToInsertForKit(kit, inboundKit, true);
+                }
+                logger.info("Finished kit " + (kitCounter+1) + " / " + kits.size() + " for participant " + participantShortId);
+            }
+        }
+    }
+
+    public static String decideWhatToInsertForKit(Kit fileKit, Kit dbKit, boolean isRetrun) {
+        fileKit.setDsmKitRequestId(dbKit.getDsmKitRequestId());
+        fileKit.setTrackingToId(dbKit.getTrackingToId());
+        fileKit.setTrackingReturnId(dbKit.getTrackingReturnId());
+        if (StringUtils.isBlank(fileKit.getShipmentId())) {
+            fileKit.setShipmentId(dbKit.getShipmentId());
+        }
+        else {
+            if (!fileKit.getShipmentId().equals(dbKit.getShipmentId())) {
+                throw new RuntimeException("Multiple shipments found for " + fileKit.getKitLabel() + " shipment ids are " + fileKit.getShipmentId() + " , " + dbKit.getShipmentId());
+            }
+        }
+        fileKit.setPackageId(dbKit.getPackageId());
+        String insertedShipmentId = null;
+        if (StringUtils.isBlank(dbKit.getLastActivityDesc())) {//insert everything if the kit doesn't have any shipments in bd
+            logger.info("Inserting Shipment, Package and all activities for kit " + fileKit.getKitLabel());
+            insertedShipmentId = insertKitInDB(fileKit, isRetrun);//most of the kits should end up here
+        }
+        else {
+            if (StringUtils.isBlank(fileKit.getPackageId()) || StringUtils.isBlank(fileKit.getShipmentId())) {
+                throw new RuntimeException("No package Id or Shipment Id was found for a kit with activity in DB! Kit label is " + fileKit.getKitLabel());
+            }
+            String firstActivityTime = isRetrun ? fileKit.getPickedUpAt() : fileKit.getShippedAt();
+            String lastActivityTime = isRetrun ? fileKit.getReceivedAt() : fileKit.getDeliveredAt();
+            if (StringUtils.isNotBlank(firstActivityTime)) {
+                if (shouldInsertBasedOnTimeForKit(dbKit, firstActivityTime)) {//last activity in DB is before the shipped at
+                    insertActivityForPackage(fileKit.getPackageId(), "Picked up", firstActivityTime);
+                }
+            }
+            if (StringUtils.isNotBlank(lastActivityTime)) {
+                if (shouldInsertBasedOnTimeForKit(dbKit, lastActivityTime)) {//last activity in DB is before the last activity in file ( we have until the transit time in db)
+                    insertActivityForPackage(fileKit.getPackageId(), "Delivered", lastActivityTime);
+                }
+            }
+        }
+        return insertedShipmentId;
+    }
+
+    public static boolean shouldInsertBasedOnTimeForKit(@NonNull Kit dbKit, String utcDateTimeString) {
```

Comment:
converts the UTC time in the CSV file and compares it with the EST time in the DB